File: filter_excel.py
import pandas as pd
import math

def wb_name():
    wb = pd.ExcelFile("download_file.xlsx")
    return wb.sheet_names


def format_day(filtered_subjects, day_mapping):
    formatted_output = ""
    for key, value in filtered_subjects.items():
        formatted_output += f"{key}:\n"
        for day_number, classes in value.items():
            formatted_output += f"{day_mapping.get(day_number, 'Неизвестный день')}:\n"
            for class_key, class_value in classes.items():
                formatted_output += f"  {class_key}. {class_value}\n"
            formatted_output += "\n"
        formatted_output += "\n"

    return formatted_output.strip()

def split_list(input_list, chunk_size=6) :
    """
    Функция разбивает список на части указанного размера и записывает результат в текстовый файл.

    Аргументы:
    - input_list (list): Исходный список для разделения.
    - chunk_size (int, по умолчанию 6): Размер каждой части списка.

    Возвращает:
    - chunks (dict): Словарь, где ключ — номер части, а значение — вложенный словарь с порядковыми номерами и элементами части.
    """
    chunks = [input_list[i :i + chunk_size] for i in range(0, len(input_list), chunk_size)]
    chunks = {i + 1 : {j + 1 : chunks[i][j] for j in range(len(chunks[i]))} for i in
              range(min(len(chunks), math.ceil(len(input_list) / chunk_size)))}

    return chunks


def filter_excel_group(file_path, sheet_name='ОСНОВНОЕ') :
    """
    Функция фильтрует данные Excel по группам, отбирая столбцы с определенными предметами.

    Аргументы:
    - file_path (str): Путь к файлу Excel.
    - sheet_name (str, по умолчанию 'ОСНОВНОЕ'): Название листа в файле.

    Возвращает:
    - group_result (dict): Словарь, где ключ — это название группы, а значение — список значений столбца.
    """
    df = pd.read_excel(file_path, sheet_name=sheet_name, engine='openpyxl', skiprows=3, nrows=36)
    subject1 = 'МДК.07.01 Управление и автоматизация баз данных\nДавыдова Л.Б.'
    subject2 = 'МДК.11.01 Технология разработки и защиты баз данных\nДавыдова Л.Б.'
    group_result = {}
    for column in df.columns :
        if subject1 in df[column].values or subject2 in df[column].values :
            group_name = df[column].to_list()[0]
            group_result[group_name] = df[column].to_list()[1 :]
    return group_result


def filter_excel(file_path, sheet_name='ОСНОВНОЕ') :
    """
    Функция загружает данные из Excel и фильтрует столбцы с указанными предметами, затем разбивает данные с помощью split_list().

    Аргументы:
    - file_path (str): Путь к файлу Excel.
    - sheet_name (str, по умолчанию 'ОСНОВНОЕ'): Название листа в файле.

    Возвращает:
    - result (dict): Словарь, где ключ — название группы, а значение — результат разбивки данных на части.
    """
    df = pd.read_excel(file_path, sheet_name=sheet_name, engine='openpyxl', skiprows=5, nrows=36)
    subject1 = 'МДК.07.01 Управление и автоматизация баз данных\nДавыдова Л.Б.'
    subject2 = 'МДК.11.01 Технология разработки и защиты баз данных\nДавыдова Л.Б.'
    flag = 0
    result = {}

    # Получаем информацию о группах
    group_data = filter_excel_group(file_path, sheet_name)

    # Проходим по каждому столбцу
    for column in df.columns :
        if subject1 in df[column].values or subject2 in df[column].values :
            sclud_list = df[column].to_list()[1 :]
            split_sclud_list = split_list(sclud_list)

            if flag < len(group_data) :
                group_name = list(group_data.keys())[flag]
                result[group_name] = split_sclud_list
                flag += 1
    return result


def filter_subjects(data, subjects) :
    """
    Фильтрует данные, оставляя только указанные предметы.

    Аргументы:
    - data (dict): Исходные данные для фильтрации.
    - subjects (list): Список предметов для фильтрации.

    Возвращает:
    - filtered_data (dict): Словарь с отфильтрованными данными по предметам.
    """
    filtered_data = {}
    for column, rows in data.items() :
        for row, cells in rows.items() :
            for cell, value in cells.items() :
                if value in subjects :
                    if column not in filtered_data :
                        filtered_data[column] = {}
                    if row not in filtered_data[column] :
                        filtered_data[column][row] = {}
                    filtered_data[column][row][cell] = value
    return filtered_data

File: filter_excel.go
package main

import (
	"fmt"
	"os"
	"github.com/tealeg/xlsx"
	"math"
)

func wbName() ([]string, error) {
	xlFile, err := xlsx.OpenFile("download_file.xlsx")
	if err != nil {
		return nil, err
	}
	var sheetNames []string
	for _, sheet := range xlFile.Sheets {
		sheetNames = append(sheetNames, sheet.Name)
	}
	return sheetNames, nil
}

func formatDay(filteredSubjects map[string]map[int]map[int]string, dayMapping map[string]string) string {
	formattedOutput := ""
	for key, value := range filteredSubjects {
		formattedOutput += fmt.Sprintf("%s:\n", key)
		for dayNumber, classes := range value {
			formattedOutput += fmt.Sprintf("%s:\n", dayMapping[fmt.Sprintf("%d", dayNumber)])
			for classKey, classValue := range classes {
				formattedOutput += fmt.Sprintf("  %d. %s\n", classKey, classValue)
			}
			formattedOutput += "\n"
		}
		formattedOutput += "\n"
	}
	return formattedOutput
}

func splitList(inputList []string, chunkSize int) map[int]map[int]string {
	chunks := make(map[int]map[int]string)
	for i := 0; i < len(inputList); i += chunkSize {
		chunk := make(map[int]string)
		for j := 0; j < chunkSize && i+j < len(inputList); j++ {
			chunk[j+1] = inputList[i+j]
		}
		chunks[i/chunkSize+1] = chunk
	}
	return chunks
}

func filterExcelGroup(filePath string, sheetName string) (map[string][]string, error) {
	xlFile, err := xlsx.OpenFile(filePath)
	if err != nil {
		return nil, err
	}
	sheet := xlFile.Sheet[sheetName]
	if sheet == nil {
		return nil, fmt.Errorf("sheet %s not found", sheetName)
	}
	subject1 := "МДК.07.01 Управление и автоматизация баз данных\nДавыдова Л.Б."
	subject2 := "МДК.11.01 Технология разработки и защиты баз данных\nДавыдова Л.Б."
	groupResult := make(map[string][]string)
	for _, row := range sheet.Rows {
		for _, cell := range row.Cells {
			text := cell.String()
			if text == subject1 || text == subject2 {
				groupName := row.Cells[0].String()
				groupResult[groupName] = append(groupResult[groupName], text)
			}
		}
	}
	return groupResult, nil
}

func filterExcel(filePath string, sheetName string) (map[string]map[int]map[int]string, error) {
	xlFile, err := xlsx.OpenFile(filePath)
	if err != nil {
		return nil, err
	}
	sheet := xlFile.Sheet[sheetName]
	if sheet == nil {
		return nil, fmt.Errorf("sheet %s not found", sheetName)
	}
	subject1 := "МДК.07.01 Управление и автоматизация баз данных\nДавыдова Л.Б."
	subject2 := "МДК.11.01 Технология разработки и защиты баз данных\nДавыдова Л.Б."
	result := make(map[string]map[int]map[int]string)
	groupData, err := filterExcelGroup(filePath, sheetName)
	if err != nil {
		return nil, err
	}
	flag := 0
	for _, row := range sheet.Rows {
		for _, cell := range row.Cells {
			text := cell.String()
			if text == subject1 || text == subject2 {
				scludList := make([]string, len(row.Cells)-1)
				for i, c := range row.Cells[1:] {
					scludList[i] = c.String()
				}
				splitScludList := splitList(scludList, 6)
				if flag < len(groupData) {
					groupName := ""
					for k := range groupData {
						groupName = k
						break
					}
					result[groupName] = splitScludList
					flag++
				}
			}
		}
	}
	return result, nil
}

func filterSubjects(data map[string]map[int]map[int]string, subjects []string) map[string]map[int]map[int]string {
	filteredData := make(map[string]map[int]map[int]string)
	for column, rows := range data {
		for row, cells := range rows {
			for cell, value := range cells {
				for _, subject := range subjects {
					if value == subject {
						if _, ok := filteredData[column]; !ok {
							filteredData[column] = make(map[int]map[int]string)
						}
						if _, ok := filteredData[column][row]; !ok {
							filteredData[column][row] = make(map[int]string)
						}
						filteredData[column][row][cell] = value
					}
				}
			}
		}
	}
	return filteredData
}
